From 11167fc5621cd1acc61c00a28e46c007ffd40b10 Mon Sep 17 00:00:00 2001
From: Yu-Ming Chang <yumin686@andestech.com>
Date: Thu, 18 Apr 2024 13:36:39 +0800
Subject: [PATCH 225/241] target/riscv: Fix HSP cosim issues. (#138)

The previous HSP patch modified dest_gpr() and also incured side effects.
This patch revert the dest_gpr() and only modify cosim related instructions.

Signed-off-by: Yu-Ming Chang <yumin686@andestech.com>

Reviewed-on: https://gitea.andestech.com/RD-SW/qemu/pulls/138
Reviewed-by: Alvin Che-Chia Chang <alvinga@andestech.com>
Co-authored-by: Yu-Ming Chang <yumin686@andestech.com>
Co-committed-by: Yu-Ming Chang <yumin686@andestech.com>
---
 target/riscv/insn_trans/trans_rvi.c.inc |  3 ++-
 target/riscv/translate.c                | 12 ++----------
 2 files changed, 4 insertions(+), 11 deletions(-)

diff --git a/target/riscv/insn_trans/trans_rvi.c.inc b/target/riscv/insn_trans/trans_rvi.c.inc
index faf6d65064..fd151c45fe 100644
--- a/target/riscv/insn_trans/trans_rvi.c.inc
+++ b/target/riscv/insn_trans/trans_rvi.c.inc
@@ -38,7 +38,8 @@ static bool trans_lui(DisasContext *ctx, arg_lui *a)
 
 static bool trans_auipc(DisasContext *ctx, arg_auipc *a)
 {
-    TCGv target_pc = dest_gpr(ctx, a->rd);
+    /* andes hsp: use tmep to make sure update will occur in gen_set_gpr() */
+    TCGv target_pc = tcg_temp_new();
     gen_pc_plus_diff(target_pc, ctx, a->imm);
     gen_set_gpr(ctx, a->rd, target_pc);
     return true;
diff --git a/target/riscv/translate.c b/target/riscv/translate.c
index 0e47d903d7..161a22f1ba 100644
--- a/target/riscv/translate.c
+++ b/target/riscv/translate.c
@@ -366,15 +366,6 @@ static TCGv dest_gpr(DisasContext *ctx, int reg_num)
     if (reg_num == 0 || get_olen(ctx) < TARGET_LONG_BITS) {
         return tcg_temp_new();
     }
-    /*
-     * andes hardware stack protection(hsp) will check any sp update.
-     * if an overflow or underflow exception occurs, sp should NOT be updated.
-     * so we retrun a temp variable to make sure the update will NOT occur
-     * before gen_set_gpr().
-     */
-    if (reg_num == 2 && ctx->cfg_ptr->ext_XAndesV5Ops) {
-        return tcg_temp_new();
-    }
     return cpu_gpr[reg_num];
 }
 
@@ -911,7 +902,8 @@ static bool gen_arith(DisasContext *ctx, arg_r *a, DisasExtend ext,
                       void (*func)(TCGv, TCGv, TCGv),
                       void (*f128)(TCGv, TCGv, TCGv, TCGv, TCGv, TCGv))
 {
-    TCGv dest = dest_gpr(ctx, a->rd);
+    /* andes hsp: use tmep to make sure update will occur in gen_set_gpr() */
+    TCGv dest = tcg_temp_new();
     TCGv src1 = get_gpr(ctx, a->rs1, ext);
     TCGv src2 = get_gpr(ctx, a->rs2, ext);
 
-- 
2.34.1

